### PR TITLE
Add RN example app (Connection Change and Error Handling)

### LINKED ISSUE #5824: Example in RN: Error handling, sync connection state change, client reset

### PR DESCRIPTION
**Pull request by elle-j**
## What, How & Why?

Adds a React Native example app for detecting changes in connection state, user state, sync errors, and product inventory, along with the backend App Services App for users who prefer to set it up via the Realm CLI.

For more information and an overview of the project structure, see the [README](https://github.com/realm/realm-js/blob/lj/example-rn-connection-and-error/examples/rn-connection-and-error/README.md).

### Main Files to Review
* `examples/rn-connection-and-error/`
  * `README.md`
  * `backend/functions/`
  * `frontend/app/`

This closes #5824 

## ☑️ ToDos
--

### CODE REVIEW

**Review by kneth — APPROVED** [review]
It is a good demo on how connection changes and errors can be handled. I really like all the comments in the code so our users easily can follow what the code is doing.

File: examples/rn-connection-and-error/frontend/app/hooks/useDemoSyncTriggers.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+  /**
+   * Trigger the client reset listeners by calling a custom Atlas Function
+   * (see `backend/functions/triggerClientReset.js`) that deletes the client
+   * files for the current user.
+   *
+   * @note
+   * This should NOT be used in production.
+   */
+  const triggerClientReset = useCallback(async () => {
+    await currentUser.functions.triggerClientReset();
```
Comment (takameyer) [inline examples/rn-connection-and-error/frontend/app/hooks/useDemoSyncTriggers.ts:134]: Seems the only way to recover from this is to restart the application.  Is that correct?

File: examples/rn-connection-and-error/frontend/app/hooks/useDemoSyncTriggers.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  const refreshAccessToken = useCallback(async () => {
+    await currentUser.refreshCustomData();
+  }, [currentUser]);
+
+  useEffect(() => {
+    /**
+     * The user listener - Will be invoked on various user related events including
+     * refresh of auth token, refresh token, custom user data, removal, and logout.
+     */
+    const handleUserEventChange = () => {
```
Comment (takameyer) [inline examples/rn-connection-and-error/frontend/app/hooks/useDemoSyncTriggers.ts:155]: This is cool.  I wonder if we could include this in the actual realm source for user events?

**Review by takameyer — APPROVED** [review]
Was able to get this up an running with no problems!  Great demo 🎉 
Code is well written and easy to parse.  Awesome job 🥳

File: examples/rn-connection-and-error/frontend/app/hooks/useDemoSyncTriggers.ts (hunk 1 above)
Comment (elle-j) [reply examples/rn-connection-and-error/frontend/app/hooks/useDemoSyncTriggers.ts:134]: Either restart or reregister as the trigger function deletes the client files. (Although we still gotta look into our `onBefore` and `onAfter` callbacks not being invoked. Will create a ticket. 👍)

File: examples/rn-connection-and-error/frontend/app/hooks/useDemoSyncTriggers.ts (hunk 2 above)
Comment (elle-j) [reply examples/rn-connection-and-error/frontend/app/hooks/useDemoSyncTriggers.ts:155]: Yeah it's neat, I believe a similar listener can also go on the `Realm.App`. Can you explain more what you mean by including it in the realm source?

**Pull request merged**